### PR TITLE
Leaderboard

### LINKED ISSUE #97: Leaderboard
Columns for leaderboard:

Rank | Name | Number of x's y'd | Email | Flagged Status
| --- | --- | --- | --- | --- |
1 | Jerome | 50 | jeromoe@example.com | approved
2 | Hils | 35 | hils@clinton.com | approved
3 | Bernie | 20 | bernie@sanders.com | approved

Sort by number of x's y'd.

If there is a pending file highlight the row in yellow, only show approved if there are NO PENDING files

Drop phone number

### PR DESCRIPTION
**Pull request by deadlybutter**
#### What's this PR do?
- Adds create leaderboard function
- Adds limit functionality (Button + Request query) 
- Updates how the flagged variable on reportbacks is assigned
- Adds unique ID migration  

#### How should this be manually tested?
Go to a competitions page, does the leaderboard look right?
Does the show all button work?
Did the migration run correctly?

#### Any background context you want to provide?
Added limits because this shit is _slow_
Added the unique ID's in migration to help speed this up per Andreas recommendation
Reportback flagged assignment changed based on yesterdays slack discussion somewhere
The rank logic is sorta complicated cause of #sports  

#### What are the relevant tickets?
Fixes #97

### CODE REVIEW

File: database/migrations/2016_03_24_153040_add_index.php [hunk 1]
```diff
@@ -0,0 +1,46 @@
+<?php
+
+use Illuminate\Database\Migrations\Migration;
+
+class AddIndex extends Migration
```
Comment (angaither) [inline database/migrations/2016_03_24_153040_add_index.php]: can you add table name to this file/class

File: app/Services/Manager.php [hunk 2]
```diff
@@ -82,6 +82,74 @@ public function exportCSV($model, $hasReportback = false)
...
+                $quantity = $reportback->quantity;
+                $flagged = $reportback->flagged;
+            }
+
+            // Push all of the data to a larger array
+            array_push($rows, ['user' => $user, 'quantity' => $quantity, 'flagged' => $flagged]);
+        }
+
+        // Sort all of the rows based on total quantity
+        usort($rows, function ($a, $b) {
```
Comment (angaither) [inline app/Services/Manager.php]: why don't we add sort to the query?

Comment (deadlybutter) [inline app/Services/Manager.php]: yeah we can do that, would you prefer its not sorted by default?

File: app/Services/Manager.php [hunk 3]
```diff
@@ -149,7 +217,13 @@ public function getUserActivity($id, $model)
             $signup->reportback->updated_at = $signup->reportback->updated_at->format('Y-m-d');
 
             // Return set flagged status to 'pending' if it is false.
-            $signup->reportback->flagged = ($signup->reportback->flagged) ? $signup->reportback->flagged : 'pending';
+            if (! isset($signup->reportback->flagged)) {
```
Comment (angaither) [inline app/Services/Manager.php]: moar comments :wink:

Comment (deadlybutter) [inline app/Services/Manager.php]: table flip

File: app/Services/Manager.php [hunk 4]
```diff
@@ -82,6 +82,74 @@ public function exportCSV($model, $hasReportback = false)
...
+                    $rank += $increment;
+                    $increment = 1;
+                }
+            }
+
+            // Assign rank to the leaderboard
+            $leaderboard[$index]['rank'] = $rank;
+        }
+
+        return $leaderboard;
```
Comment (weerd) [inline app/Services/Manager.php]: Any chance we could move away from Drupal style long-ass crazy functions and break this shiz up a bit into callable methods?

File: app/Http/Controllers/CompetitionsController.php [hunk 5]
```diff
@@ -42,11 +43,27 @@ public function __construct(UserRepositoryContract $repository, Manager $manager
...
      */
-    public function show(Competition $competition)
+    public function show(Competition $competition, Request $request)
     {
         $contest = Contest::find($competition->contest_id);
 
-        $users = $this->repository->getAll($competition->users->pluck('id')->toArray());
+        // Determine the amount of users to show in the leaderboard
+        $limit = 10;
+        $limitQuery = $request->input('limit');
```
Comment (weerd) [inline app/Http/Controllers/CompetitionsController.php]: Is this coming in from a URL param?

File: app/Services/Manager.php (hunk 4 above)
Comment (deadlybutter) [inline app/Services/Manager.php]: we're moving some of this functionality into drupal now so hang tight

File: app/Http/Controllers/CompetitionsController.php (hunk 5 above)
Comment (deadlybutter) [inline app/Http/Controllers/CompetitionsController.php]: yes `/competitions/8?limit=all`

File: app/Services/Manager.php (hunk 4 above)
Comment (weerd) [inline app/Services/Manager.php]: Oh! Hokay. Sorry, playing catch up post React class.

**Pull request closed without merging**